lib/Dialect/TTNN/Transforms/Workarounds/TTNNWorkarounds.cpp

```diff
@@ -478,44 +479,45 @@ class TTNNWorkarounds : public impl::TTNNWorkaroundsBase<TTNNWorkarounds> {
-                                              config))) {
-        signalPassFailure();
-        return;
-      }
+      patterns.add<
+          TTNNAllReduceWorkarounds,
+          workarounds::decomposition::ReduceOpsRewritePattern<ttnn::SumOp>,
+          workarounds::decomposition::ReduceOpsRewritePattern<ttnn::MaxOp>,
+          workarounds::decomposition::ReduceOpsRewritePattern<ttnn::MeanOp>>(
+          &getContext());
+
+      runRewritePatterns(std::move(patterns));
     }
     if (layouotWorkaroundsEnabled) {
       RewritePatternSet patterns(&getContext());
       patterns.add<TTNNOperandsWorkaroundsRewriter>(&getContext());
 
-      FrozenRewritePatternSet patternSet(std::move(patterns));
-      GreedyRewriteConfig config = GreedyRewriteConfig();
-      // This configuration specifies that the rewriter should traverse the IR
-      // in a top-down order.
-      config.useTopDownTraversal = true;
-      // This configuration specifies the maximum number of iterations the
-      // rewriter will perform on the IR. The rewriter will iterate through the
-      // IR until a fixpoint is reached. All workarounds should be applied
-      // during the first iteration. If the workarounds are not applied in the
-      // first iteration, it indicates a bug in the workarounds implementation.
-      // Although the workarounds are applied in the first iteration, the
-      // rewriter must iterate through the IR once more to confirm that the
-      // fixpoint is reached. If the fixpoint is not reached in the second
-      // iteration, it indicates a bug in the workarounds implementation.
-      config.maxIterations = 2;
-      if (failed(applyPatternsAndFoldGreedily(getOperation(), patternSet,
-                                              config))) {
-        signalPassFailure();
-        return;
-      }
+      runRewritePatterns(std::move(patterns));
+    }
+  }
+
+private:
+  void runRewritePatterns(RewritePatternSet &&patterns) {
+    FrozenRewritePatternSet patternSet(std::move(patterns));
+    GreedyRewriteConfig config = GreedyRewriteConfig();
+    // This configuration specifies that the rewriter should traverse the IR
+    // in a top-down order.
+    config.useTopDownTraversal = true;
+    // This configuration specifies the maximum number of iterations the
+    // rewriter will perform on the IR. The rewriter will iterate through the
+    // IR until a fixpoint is reached. All workarounds should be applied
+    // during the first iteration. If the workarounds are not applied in the
+    // first iteration, it indicates a bug in the workarounds implementation.
+    // Although the workarounds are applied in the first iteration, the
+    // rewriter must iterate through the IR once more to confirm that the
+    // fixpoint is reached. If the fixpoint is not reached in the second
+    // iteration, it indicates a bug in the workarounds implementation.
+    config.maxIterations = 2;
```

Comment:
Are we aligned on this setting for workaround decompositions? Unlike the layout rewriter, which guarantees completion in a maximum of 2 iterations, decomposition rewrites may require more iterations. This is because a pattern can decompose into sub-patterns that also require rewrites, making it harder to reach a fixpoint in a limited number of iterations. This means that if the fixpoint isn't reached in 2 iterrations, we will have not fully applied workaround decompositions which is probably invalid IR in our case. We should consider setting a higher value or even kNoLimit for decomposition workarounds.

Reply:
I can add additional argument to `runRewritePatterns` to specify the number of iterations. The problem with `kNoLimit` is that it goes into infinite loop trying to rewrite the Reduce op all over again, even though I return failure if the workaround is already applied, I didn't know how to stop it other than limiting the number of iterations. Do you have some idea?

Reply:
GreedyRewriter continues operating until either a fixpoint is reached or the specified number of iterations is completed. If the maximum number of iterations is set to kNoLimit, there is a possibility of encountering an infinite loop. This could indicate that the rewrite pattern itself is generating the infinite loop. I have thoroughly reviewed the pattern and found that it should not create an infinite loop. Do you have a specific example of a failure that occurred when the maximum iterations were set to kNoLimit?

Reply:
Tried it again now and no hangs this time, not sure what was the issue back then, maybe the build somehow got in inconsistent state... I've added a parameter for `maxIterations` into `runRewritePatterns` and changed the value back to `GreedyRewriteConfig::kNoLimit` for decomposition workarounds.